mitxgraders/sampling.py

```diff
@@ -338,6 +349,85 @@ class RealMatrices(RealMathArrays):
     })
 
 
+class AbstractSquareMatrices(VariableSamplingSet):  # pylint: disable=abstract-method
+    """
+    Abstract class representing a collection of square matrices
+
+    Config:
+    =======
+        dimension: Positive integer that specifies the dimension of the matrix (default 2)
+    """
+
+    # This is an abstract base class
+    __metaclass__ = abc.ABCMeta
+
+    # Store the dimension of the square matrix
+    schema_config = Schema({
+        Required('dimension', default=2): Positive(int)
+    })
+
+
+class IdentityMatrixMultiples(AbstractSquareMatrices):
+    """
+    Class representing a collection of multiples of the identity matrix
+    of a given dimension
+
+    Config:
+    =======
+        dimension: Positive integer that specifies the dimension of the matrix (default 2)
+        sampler: A scalar sampling set for the multiplicative constant (default [1, 5])
+
+    Usage:
+    ======
+
+    By default, we generate 2x2 matrices:
+    >>> matrices = IdentityMatrixMultiples()
+    >>> matrices.gen_sample().shape
+    (2, 2)
+
+    We can generate NxN matrices by specifying the dimension:
+    >>> matrices = IdentityMatrixMultiples(dimension=4)
+    >>> matrices.gen_sample().shape
+    (4, 4)
+
+    The scalar multiple can be generated in a number of ways:
+    >>> matrices = IdentityMatrixMultiples(sampler=[1,3])
+    >>> matrices = IdentityMatrixMultiples(sampler=(1, 3, 5))
+    >>> sect = ComplexSector(modulus=[0,1], argument=[-np.pi,np.pi])
+    >>> matrices = IdentityMatrixMultiples(sampler=sect)
+
+    The resulting samples are simply a scalar times the identity matrix:
+    >>> matrices = IdentityMatrixMultiples()
+    >>> m = matrices.gen_sample()
+    >>> m == m[0, 0] * np.eye(2)
+    array([[ True,  True],
+           [ True,  True]], dtype=bool)
+    """
+
+    # Sampling set for the multiplicative constant
+    # Accept anything that FormulaGrader would accept for a sampling set, restricted to
+    # scalar sampling sets. Hence, ScalarSamplingSets, ranges, and objects that can be
+    # coerced into discrete sets are allowed.
+    # Note: Does not support DependentSampler, as that is not guaranteed to return a
+    # scalar value.
+    schema_config = AbstractSquareMatrices.schema_config.extend({
+        Required('sampler', default=RealInterval()): Any(ScalarSamplingSet,
+                                                         All(list, Coerce(RealInterval)),
+                                                         Coerce(DiscreteSet))
+    })
+
+    def gen_sample(self):
+        """
+        Generates an identity matrix of specified dimension multiplied by a random scalar
+        """
+        # Sample the multiplicative constant
+        scaling = self.config['sampler'].gen_sample()
+        # Create the matrix
+        matrix = scaling * np.eye(self.config['dimension'])
```

Comment:
Suggestion: rename this `array` or something, since it's not a MathArray Matrix.

Reply:
Ok, can do.